### PR TITLE
initial observation block data insert

### PR DESCRIPTION
**Pull request by anissa-agahchen**
This PR contains:
* Database model for block observation
* wiring up the front-end to send a request payload
* set up the API endpoint to generically accept block observations, and other types of observations in the future
* tests

### CODE REVIEW

File: api/src/paths/project/{projectId}/survey/{surveyId}/observations/create.ts [hunk 1]
```diff
@@ -0,0 +1,152 @@
...
+
+/**
+ * Creates a new block observation record.
+ *
+ * @returns {RequestHandler}
+ */
+export function createBlockObservation(): RequestHandler {
+  return async (req, res) => {
+    defaultLog.debug({
+      label: 'Add a block observation to a survey',
```
Comment (NickPhura) [inline api/src/paths/project/{projectId}/survey/{surveyId}/observations/create.ts]: nit: can we make the label 'createBlockObservation'

File: app/src/features/observations/BlockObservationPage.tsx [hunk 2]
```diff
@@ -17,6 +17,13 @@ import { useBiohubApi } from 'hooks/useBioHubApi';
 import { IGetProjectForViewResponse } from 'interfaces/useProjectApi.interface';
 import CircularProgress from '@material-ui/core/CircularProgress';
 import { IGetSurveyForViewResponse } from 'interfaces/useSurveyApi.interface';
+import { ICreateBlockObservationPostRequest } from 'interfaces/useObservationApi.interface';
+import moment from 'moment';
+import { APIError } from 'hooks/api/useAxios';
+import {
+  //ErrorDialog,
```
Comment (NickPhura) [inline app/src/features/observations/BlockObservationPage.tsx]: nit: remove if unused

File: app/src/features/observations/BlockObservationPage.tsx [hunk 3]
```diff
@@ -140,6 +160,41 @@ const BlockObservationPage = () => {
...
+      }
+    };
+
+    try {
+      const response = await biohubApi.observation.createBlockObservation(projectId, surveyId, postData);
+
+      if (!response) {
+        return;
+      } else {
+        console.log('we have a successful entry');
```
Comment (NickPhura) [inline app/src/features/observations/BlockObservationPage.tsx]: nit: remove

File: app/src/features/observations/BlockObservationPage.tsx [hunk 4]
```diff
@@ -140,6 +160,41 @@ const BlockObservationPage = () => {
...
+        setEnableCancelCheck(false);
+        history.push(`/projects/${projectId}/surveys/${surveyId}/observations`);
+      }
+    } catch (error) {
+      console.log('there is an error');
```
Comment (NickPhura) [inline app/src/features/observations/BlockObservationPage.tsx]: nit: remove

File: api/src/openapi/schemas/observation.ts [hunk 5]
```diff
@@ -1,21 +1,13 @@
 /**
- * Basic response object for an observation.
+ * Basic response object for a survey.
```
Comment (sdevalapurkar) [inline api/src/openapi/schemas/observation.ts]: why did this change... it should say observation right?

Comment (sdevalapurkar) [reply api/src/openapi/schemas/observation.ts]: I dont think this model should change at all - it needs to return blocks

File: api/src/queries/observation/observation-create-queries.ts [hunk 6]
```diff
@@ -0,0 +1,53 @@
+import SQL, { SQLStatement } from 'sql-template-strings';
+import { getLogger } from '../../utils/logger';
+import { PostBlockObservationObject } from '../../models/block-observation-create';
+
+const defaultLog = getLogger('queries/observation-create-queries');
+
+/**
+ * SQL query to insert a row in the webform_draft table.
```
Comment (sdevalapurkar) [inline api/src/queries/observation/observation-create-queries.ts]: copy paste error for this comment

File: api/src/queries/observation/observation-create-queries.ts [hunk 7]
```diff
@@ -0,0 +1,53 @@
...
+ *
+ * @param {number} systemUserId the ID of the user in context
+ * @param {string} name name of the draft record
+ * @param {unknown} data JSON data blob
```
Comment (sdevalapurkar) [inline api/src/queries/observation/observation-create-queries.ts]: same here

File: app/src/features/observations/BlockObservationPage.tsx [hunk 8]
```diff
@@ -140,6 +160,41 @@ const BlockObservationPage = () => {
     return <CircularProgress className="pageProgress" size={40} />;
   }
 
+  const handleSaveAndExit = async (values: ICreateBlockObservationPostRequest) => {
+    if (!formikRef?.current) {
+      return;
+    }
+
```
Comment (sdevalapurkar) [inline app/src/features/observations/BlockObservationPage.tsx:196]: here we need to do a formikRef.current.handleSubmit or something to ensure form fields are all valid and no yup errors I think

File: app/src/features/observations/BlockObservationPage.tsx [hunk 9]
```diff
@@ -140,6 +160,41 @@ const BlockObservationPage = () => {
...
+    const postData: any = {
+      block_name: formikRef.current.values.block_name,
+      start_datetime: moment(formikRef.current.values.date + ' ' + formikRef.current.values.start_time).toISOString(),
+      end_datetime: moment(formikRef.current.values.date + ' ' + formikRef.current.values.end_time).toISOString(),
```
Comment (sdevalapurkar) [inline app/src/features/observations/BlockObservationPage.tsx]: can we use template strings here to construct this

**Review by sdevalapurkar — CHANGES_REQUESTED** [review]

File: app/src/interfaces/useObservationApi.interface.ts [hunk 10]
```diff
@@ -37,3 +37,64 @@ export interface IGetObservationResponse {
...
+ *
+ * @export
+ * @interface ICreateBlockObservationPostRequest
+ */
+export interface ICreateBlockObservationPostRequest {
+  metaData: IBlockObservationMeta;
+  tableData: IBlockObservationTableData;
+}
+
+export interface IBlockObservationMeta {
```
Comment (sdevalapurkar) [inline app/src/interfaces/useObservationApi.interface.ts]: you dont need this
we can just use `IBlockObservationForm` which is defined in BlockObservationForm.tsx

**Review by sdevalapurkar — CHANGES_REQUESTED** [review]

File: api/src/paths/project/{projectId}/survey/{surveyId}/observations/create.ts [hunk 11]
```diff
@@ -0,0 +1,152 @@
...
+
+const defaultLog = getLogger('paths/project/{projectId}/survey/{surveyId}/observations/create');
+
+export const POST: Operation = [
+  logRequest('paths/project/{projectId}/survey/{surveyId}/observations/create', 'POST'),
+  createBlockObservation()
+];
+
+POST.apiDoc = {
+  description: 'Create a new block observation.',
```
Comment (sdevalapurkar) [inline api/src/paths/project/{projectId}/survey/{surveyId}/observations/create.ts]: this is not the correct description - this endpoint is generic for all observation creates not just block type

File: api/src/paths/project/{projectId}/survey/{surveyId}/observations/create.ts [hunk 12]
```diff
@@ -0,0 +1,152 @@
...
+  tags: ['block observation'],
```
Comment (sdevalapurkar) [inline api/src/paths/project/{projectId}/survey/{surveyId}/observations/create.ts]:
```suggestion
  tags: ['observation'],
```

File: api/src/paths/project/{projectId}/survey/{surveyId}/observations/create.ts [hunk 13]
```diff
@@ -0,0 +1,152 @@
...
+  security: [
+    {
+      Bearer: [SYSTEM_ROLE.SYSTEM_ADMIN, SYSTEM_ROLE.PROJECT_ADMIN]
+    }
+  ],
+  requestBody: {
+    description: 'Block Observation post request object.',
```
Comment (sdevalapurkar) [inline api/src/paths/project/{projectId}/survey/{surveyId}/observations/create.ts]:
```suggestion
    description: 'Observation post request object.',
```

File: api/src/paths/project/{projectId}/survey/{surveyId}/observations/create.ts [hunk 14]
```diff
@@ -0,0 +1,152 @@
...
+    content: {
+      'application/json': {
+        schema: {
+          title: 'Block Observation request object',
```
Comment (sdevalapurkar) [inline api/src/paths/project/{projectId}/survey/{surveyId}/observations/create.ts]:
```suggestion
          title: 'Observation request object',
```

File: api/src/paths/project/{projectId}/survey/{surveyId}/observations/create.ts [hunk 15]
```diff
@@ -0,0 +1,152 @@
...
+      }
+    }
+  },
+  responses: {
+    200: {
+      description: 'Block Observation response object.',
+      content: {
+        'application/json': {
+          schema: {
+            ...(blockObservationIdResponseObject as object)
```
Comment (sdevalapurkar) [inline api/src/paths/project/{projectId}/survey/{surveyId}/observations/create.ts]: this entire endpoint needs to change to accept the entity type and then handle the insertion of block observation if type is block - cannot be only for block but instead needs to be generic

**Review by sdevalapurkar — CHANGES_REQUESTED** [review]

File: app/src/features/observations/BlockObservationPage.tsx (hunk 8 above)
Comment (sdevalapurkar) [reply app/src/features/observations/BlockObservationPage.tsx:196]: await formikRef.current?.submitForm();

Comment (sdevalapurkar) [reply app/src/features/observations/BlockObservationPage.tsx:196]:
```
await formikRef.current?.submitForm();

    const isValid = await validateFormFieldsAndReportCompletion(
      formikRef.current?.values,
      formikRef.current?.validateForm
    );

    if (!isValid) {
      showCreateErrorDialog({
        dialogTitle: 'Add Observation Form Incomplete',
        dialogText:
          'The form is missing some required fields/sections highlighted in red. Please fill them out and try again.'
      });

      return;
    }
```

File: app/src/features/observations/BlockObservationPage.tsx (hunk 9 above)
Comment (sdevalapurkar) [reply app/src/features/observations/BlockObservationPage.tsx]: moment(`${formikRef.current.values.date} ${formikRef.current.values.start_time}`).toISOString()

File: api/src/models/block-observation-create.ts [hunk 16]
```diff
@@ -0,0 +1,22 @@
...
+ */
+export class PostBlockObservationObject {
+  block_name: number;
+  start_datetime: string;
+  end_datetime: string;
+  observation_count: number;
+  observation_data: string;
+
+  constructor(obj?: any) {
+    //this.block_name = (obj?.block_name && Number(obj?.block_name)) || ((null as unknown) as number);
```
Comment (sdevalapurkar) [inline api/src/models/block-observation-create.ts]: remove this line

File: api/src/openapi/schemas/observation.ts [hunk 17]
```diff
@@ -10,21 +10,13 @@ export const observationUpdateGetResponseObject = {
 /**
  * Basic response object for an observation.
  */
-export const observationIdResponseObject = {
-  title: 'Observation Response Object',
+export const observationResponseObject = {
+  title: 'observation Object',
```
Comment (sdevalapurkar) [inline api/src/openapi/schemas/observation.ts]:
```suggestion
  title: 'Observation Response Object',
```

File: api/src/openapi/schemas/observation.ts [hunk 18]
```diff
@@ -10,21 +10,13 @@ export const observationUpdateGetResponseObject = {
...
-      items: {
-        type: 'object',
-        properties: {
-          id: {
-            type: 'number'
-          }
-        }
-      }
+    id: {
+      type: 'number'
```
Comment (sdevalapurkar) [inline api/src/openapi/schemas/observation.ts]: this should not change, I dont understand why it has been modified
the endpoint is returning an object with blocks property not just an id

File: api/src/paths/project/{projectId}/survey/{surveyId}/observations/create.ts [hunk 19]
```diff
@@ -0,0 +1,143 @@
...
+    description: 'Observation post request object.',
+    content: {
+      'application/json': {
+        schema: {
+          title: 'Observation request object',
+          type: 'object',
+          required: ['observation_type'],
+          properties: {
+            observation_type: {
+              title: 'type of observation',
```
Comment (sdevalapurkar) [inline api/src/paths/project/{projectId}/survey/{surveyId}/observations/create.ts]:
```suggestion
              title: 'Type of observation (eg: block)',
```

File: api/src/paths/project/{projectId}/survey/{surveyId}/observations/create.ts [hunk 20]
```diff
@@ -0,0 +1,143 @@
...
+      }
+    }
+  },
+  responses: {
+    200: {
+      description: 'Observation response object.',
+      content: {
+        'application/json': {
+          schema: {
+            ...(observationResponseObject as object)
```
Comment (sdevalapurkar) [inline api/src/paths/project/{projectId}/survey/{surveyId}/observations/create.ts]: you need to create something new here, cannot modify the one I created for `getForUpdate` response

File: api/src/paths/project/{projectId}/survey/{surveyId}/observations/create.ts [hunk 21]
```diff
@@ -0,0 +1,143 @@
...
+      'req.body': req.body
+    });
+
+    if (!req.params.surveyId) {
+      throw new HTTP400('Missing required path param `surveyId`');
+    }
+
+    const connection = getDBConnection(req['keycloak_token']);
+
+    let sanitizedObservationData = null;
```
Comment (sdevalapurkar) [inline api/src/paths/project/{projectId}/survey/{surveyId}/observations/create.ts]:
```suggestion
    let sanitizedObservationData;
```

File: api/src/paths/project/{projectId}/survey/{surveyId}/observations/create.ts [hunk 22]
```diff
@@ -0,0 +1,143 @@
...
+    }
+
+    if (!sanitizedObservationData) {
+      throw new HTTP400('Missing observation data');
+    }
+
+    let observationSQLStatement = null;
+
+    if (req.body.observation_type === 'block') {
+      observationSQLStatement = postBlockObservationSQL(Number(req.params.surveyId), sanitizedObservationData);
```
Comment (sdevalapurkar) [inline api/src/paths/project/{projectId}/survey/{surveyId}/observations/create.ts:112]: question

can all this be done within the first if statement? why two identical if statements

File: api/src/queries/observation/observation-create-queries.ts [hunk 23]
```diff
@@ -0,0 +1,52 @@
...
+      ${observationPostData.end_datetime},
+      ${observationPostData.observation_count},
+      ${observationPostData.observation_data}
+    )
+    RETURNING
+      id
+  `;
+
+  defaultLog.debug({
+    label: 'postBlockObservationSQL',
```
Comment (sdevalapurkar) [inline api/src/queries/observation/observation-create-queries.ts:45]: lets please add a unit test to this function in this pr
it should be super easy (copy paste almost)

File: app/src/features/observations/BlockObservationPage.tsx [hunk 24]
```diff
@@ -172,6 +190,61 @@ const BlockObservationPage = () => {
...
+        }
+      }
+    };
+
+    try {
+      const response = await biohubApi.observation.createBlockObservation(projectId, surveyId, postData);
+
+      if (!response) {
+        return;
+      } else {
```
Comment (sdevalapurkar) [inline app/src/features/observations/BlockObservationPage.tsx]: dont need an else here
you are returning in the if so can remove the else

File: app/src/features/observations/BlockObservationPage.tsx [hunk 25]
```diff
@@ -172,6 +190,61 @@ const BlockObservationPage = () => {
...
+        setEnableCancelCheck(false);
+        history.push(`/projects/${projectId}/surveys/${surveyId}/observations`);
+      }
+    } catch (error) {
+      const apiError = error as APIError;
+      showErrorDialog({ dialogText: apiError.message, dialogErrorDetails: apiError.errors, open: true });
+    }
+  };
+
+  console.log(formikRef?.current?.values);
```
Comment (sdevalapurkar) [inline app/src/features/observations/BlockObservationPage.tsx]:
```suggestion
```

File: app/src/features/observations/BlockObservationPage.tsx [hunk 26]
```diff
@@ -172,6 +190,61 @@ const BlockObservationPage = () => {
     return <CircularProgress className="pageProgress" size={40} />;
   }
 
+  const handleSaveAndExit = async (values: ICreateBlockObservationPostRequest) => {
```
Comment (sdevalapurkar) [inline app/src/features/observations/BlockObservationPage.tsx]: this function does not use the argument it receives... I dont think values is ever used
can just be:
```
const handleSaveAndExit = async () => {
```

File: app/src/features/observations/BlockObservationPage.tsx [hunk 27]
```diff
@@ -233,7 +306,7 @@ const BlockObservationPage = () => {
                     variant="contained"
                     color="primary"
                     data-testid="save-and-exit-button"
-                    onClick={() => console.log('add and exit functionality')}
+                    onClick={(values: any) => handleSaveAndExit(values)}
```
Comment (sdevalapurkar) [inline app/src/features/observations/BlockObservationPage.tsx]:
```suggestion
                    onClick={handleSaveAndExit}
```

File: app/src/hooks/api/useObservationApi.test.ts [hunk 28]
```diff
@@ -41,4 +42,47 @@ describe('useObservationApi', () => {
...
+        weather_description: 't',
+        description_of_habitat: 'u',
+        aircraft_company: 'v',
+        aircraft_type: 'w',
+        aircraft_registration_number: 123,
+        aircraft_gps_model: 'y',
+        aircraft_gps_datum: 'z',
+        aircraft_gps_readout: 'zz'
+      },
+      tableData: { data: [['1', '2']] }
```
Comment (sdevalapurkar) [inline app/src/hooks/api/useObservationApi.test.ts]: instead of redefining this here, can just use what is in `observation-helpers.ts`
will save like 30 lines of code

File: app/src/hooks/api/useObservationApi.ts [hunk 29]
```diff
@@ -21,9 +26,27 @@ const useObservationApi = (axios: AxiosInstance) => {
 
     return data;
   };
+  /**
+   * Create a new block observation
+   *
+   * @param {ICreateBlockObservationRequest} blockObservation
+   * @return {*}  {Promise<ICreateBlockObservationResponse>}
+   */
+  const createBlockObservation = async (
```
Comment (sdevalapurkar) [inline app/src/hooks/api/useObservationApi.ts]: this function is not named correctly
it is now a generic endpoint so should be called `createObservation` as it deals with not only block type anymore :)

File: app/src/hooks/api/useObservationApi.ts [hunk 30]
```diff
@@ -21,9 +26,27 @@ const useObservationApi = (axios: AxiosInstance) => {
...
+      `/api/project/${projectId}/survey/${surveyId}/observations/create`,
+      blockObservation
+    );
+
+    return data;
+  };
 
   /**
-   * Get details for a single observation for update purposes.
+   *Get details for a single observation for update purposes.
```
Comment (sdevalapurkar) [inline app/src/hooks/api/useObservationApi.ts]:
```suggestion
   * Get details for a single observation for update purposes.
```

File: app/src/interfaces/useObservationApi.interface.ts [hunk 31]
```diff
@@ -37,3 +37,34 @@ export interface IGetObservationResponse {
...
+  blocks: IGetBlocksListResponse[];
+}
+
+/**
+ * Create blockObservation post object.
+ *
+ * @export
+ * @interface ICreateBlockObservationPostRequest
+ */
+export interface ICreateBlockObservationPostRequest {
```
Comment (sdevalapurkar) [inline app/src/interfaces/useObservationApi.interface.ts:60]: this does not seem correct based on the new data you are sending to the endpoint, I think it needs to be updated to match

**Review by sdevalapurkar — CHANGES_REQUESTED** [review]

File: app/src/interfaces/useObservationApi.interface.ts (hunk 31 above)
Comment (anissa-agahchen) [reply app/src/interfaces/useObservationApi.interface.ts:60]: I'm going to leave this for now .... until we know more about the requirements of the other observation types.

Comment (sdevalapurkar) [reply app/src/interfaces/useObservationApi.interface.ts:60]: no what I meant is that this doesnt match what is being sent and needs to be updated to match this structure:
```
observation_type,
observation_post_data: {
  block_name
  start_datetime
  end_datetime
  observation_count
  observation_data: {
    metaData
    tableData: {
      data
    }
  }
}
```

Comment (anissa-agahchen) [reply app/src/interfaces/useObservationApi.interface.ts:60]: oh i see!

File: api/src/openapi/schemas/observation.ts [hunk 32]
```diff
@@ -10,8 +10,8 @@ export const observationUpdateGetResponseObject = {
 /**
  * Basic response object for an observation.
  */
-export const observationResponseObject = {
-  title: 'observation Object',
+export const observationPostResponseObject = {
+  title: 'Observation Response Object',
```
Comment (sdevalapurkar) [inline api/src/openapi/schemas/observation.ts]: `title: 'Observation Post Response Object'`

File: api/src/paths/project/{projectId}/survey/{surveyId}/observations/list.ts [hunk 33]
```diff
@@ -4,7 +4,7 @@ import { SYSTEM_ROLE } from '../../../../../../constants/roles';
 import { getDBConnection } from '../../../../../../database/db';
 import { HTTP400 } from '../../../../../../errors/CustomError';
 import { GetBlockObservationListData } from '../../../../../../models/observation-view';
-import { observationResponseObject } from '../../../../../../openapi/schemas/observation';
+import { observationPostResponseObject } from '../../../../../../openapi/schemas/observation';
```
Comment (sdevalapurkar) [inline api/src/paths/project/{projectId}/survey/{surveyId}/observations/list.ts]: this should be `observationGetResponseObject` not the post one

File: api/src/paths/project/{projectId}/survey/{surveyId}/observations/list.ts [hunk 34]
```diff
@@ -50,7 +50,7 @@ GET.apiDoc = {
           schema: {
             type: 'array',
             items: {
-              ...(observationResponseObject as object)
+              ...(observationPostResponseObject as object)
```
Comment (sdevalapurkar) [inline api/src/paths/project/{projectId}/survey/{surveyId}/observations/list.ts]: same here

File: api/src/openapi/schemas/observation.ts [hunk 35]
```diff
@@ -10,8 +10,22 @@ export const observationUpdateGetResponseObject = {
...
+  required: ['id'],
+  properties: {
+    id: {
+      type: 'number'
+    }
+  }
+};
+
+/**
+ * Basic response object for an observation.
```
Comment (sdevalapurkar) [inline api/src/openapi/schemas/observation.ts]:
```suggestion
 * Basic get response object for an observation.
```

File: api/src/openapi/schemas/observation.ts [hunk 36]
```diff
@@ -10,8 +10,22 @@ export const observationUpdateGetResponseObject = {
 /**
  * Basic response object for an observation.
```
Comment (sdevalapurkar) [inline api/src/openapi/schemas/observation.ts]:
```suggestion
 * Basic post response object for an observation.
```

File: api/src/paths/project/{projectId}/survey/{surveyId}/observations/create.ts [hunk 37]
```diff
@@ -0,0 +1,143 @@
...
+      sanitizedObservationData = (req.body && new PostBlockObservationObject(req.body.observation_post_data)) || null;
+    }
+
+    if (!sanitizedObservationData) {
+      throw new HTTP400('Missing observation data');
+    }
+
+    let observationSQLStatement = null;
+
+    if (req.body.observation_type === 'block') {
```
Comment (sdevalapurkar) [inline api/src/paths/project/{projectId}/survey/{surveyId}/observations/create.ts:110]: so is it possible to not duplicate these identical if statements somehow? like put all the logic into one conditional?

File: app/src/hooks/api/useObservationApi.ts [hunk 38]
```diff
@@ -21,6 +26,22 @@ const useObservationApi = (axios: AxiosInstance) => {
 
     return data;
   };
+  /**
+   * Create a new block observation
```
Comment (sdevalapurkar) [inline app/src/hooks/api/useObservationApi.ts:30]:
```suggestion
   * Create a new observation
```

File: app/src/interfaces/useObservationApi.interface.ts [hunk 39]
```diff
@@ -55,16 +55,28 @@ export interface IGetObservationsListResponse {
...
 }
 
-export interface ICreateBlockObservationPostResponse {
+export interface ICreateObservationPostResponse {
   id: number;
 }
+
+export interface ICreateObservationPostRequest {
+  observation_type: string;
+  observation_post_data: ICreateBlockObservationPostRequest;
```
Comment (sdevalapurkar) [inline app/src/interfaces/useObservationApi.interface.ts:81]: nice! 👍🏼

File: api/src/paths/project/{projectId}/survey/{surveyId}/observations/create.ts (hunk 37 above)
Comment (anissa-agahchen) [reply api/src/paths/project/{projectId}/survey/{surveyId}/observations/create.ts:110]: yeah .. was talking to Nick earlier.  There is ... but might be good to wait until we know more about the requirements of the other observations before doing that ... just in case this whole approach has to be scrapped

Comment (sdevalapurkar) [reply api/src/paths/project/{projectId}/survey/{surveyId}/observations/create.ts:110]: ok yeah that sounds good then

File: api/src/queries/observation/observation-create-queries.test.ts [hunk 40]
```diff
@@ -0,0 +1,32 @@
...
+  postBlockObservationSQL
+} from './observation-create-queries';
+
+describe('postBlockObservationSQL', () => {
+  it('returns null when null surveyId param provided', () => {
+    const observation = new PostBlockObservationObject();
+    const response = postBlockObservationSQL((null as unknown) as number, observation);
+
+    expect(response).to.be.null;
+  });
```
Comment (sdevalapurkar) [inline api/src/queries/observation/observation-create-queries.test.ts:12]: could add a test case for when surveyId is not null but data is null

File: api/src/paths/project/{projectId}/survey/{surveyId}/observations/create.test.ts [hunk 41]
```diff
@@ -0,0 +1,207 @@
...
+      const result = create.createObservation();
+
+      await result(
+        { ...sampleReq, body: { observation_type: null, observation_post_data: sampleReq.body.observation_post_data } },
+        (null as unknown) as any,
+        (null as unknown) as any
+      );
+      expect.fail();
+    } catch (actualError) {
+      console.log('actualError', actualError);
```
Comment (sdevalapurkar) [inline api/src/paths/project/{projectId}/survey/{surveyId}/observations/create.test.ts]:
```suggestion
```

File: api/src/paths/project/{projectId}/survey/{surveyId}/observations/create.test.ts [hunk 42]
```diff
@@ -0,0 +1,207 @@
...
+      const result = create.createObservation();
+
+      await result(
+        { ...sampleReq, body: { observation_type: 'block', observation_post_data: null } },
+        (null as unknown) as any,
+        (null as unknown) as any
+      );
+      expect.fail();
+    } catch (actualError) {
+      console.log('actualError', actualError);
```
Comment (sdevalapurkar) [inline api/src/paths/project/{projectId}/survey/{surveyId}/observations/create.test.ts]:
```suggestion
```

File: api/src/paths/project/{projectId}/survey/{surveyId}/observations/create.test.ts [hunk 43]
```diff
@@ -0,0 +1,207 @@
...
+      systemUserId: () => {
+        return 20;
+      }
+    });
+
+    try {
+      const result = create.createObservation();
+
+      await result(
+        { ...sampleReq, body: { observation_type: null, observation_post_data: sampleReq.body.observation_post_data } },
```
Comment (sdevalapurkar) [inline api/src/paths/project/{projectId}/survey/{surveyId}/observations/create.test.ts]: if you want you can also just say:
```suggestion
        { ...sampleReq, body: { ...sampleReq.body, observation_type: null } },
```

Comment (anissa-agahchen) [reply api/src/paths/project/{projectId}/survey/{surveyId}/observations/create.test.ts]: nice, thanks!

File: api/src/queries/observation/observation-create-queries.test.ts [hunk 44]
```diff
@@ -0,0 +1,37 @@
...
+    expect(response).to.not.be.null;
+  });
+
+  it('returns null when surveyId is valid but data is null', () => {
+    const blockObservationData = null;
+
+    const postBlockObservationObject = new PostBlockObservationObject(blockObservationData);
+    const response = postBlockObservationSQL(1, postBlockObservationObject);
+
+    expect(response).to.not.be.null;
```
Comment (sdevalapurkar) [inline api/src/queries/observation/observation-create-queries.test.ts]:
```suggestion
    const response = postBlockObservationSQL(1, null);

    expect(response).to.be.null;
```

Comment (sdevalapurkar) [reply api/src/queries/observation/observation-create-queries.test.ts]:
```
const x = new PostBlockObservationObject(null);
```
will return:
```
x = {
    block_name: null,
    blah: null,
    blah2: null
    ...
}
```

**Review by sdevalapurkar — DISMISSED** [review]
LGTM, sweet 🥇

**Review by sdevalapurkar — APPROVED** [review]
